llmfoundry/data/text_data.py

```diff
@@ -293,6 +295,47 @@ def build_text_dataloader(
         timeout=cfg.get('timeout', 0),
     )
 
+    # If we pretokenized, we may not have padding, in which the
+    # tokenizer may not have a pad_token_id. In this case, we can
+    # just use the default token counting function.
+    token_counting_func = None
+    if tokenizer.pad_token_id is not None:
+        token_counting_func = get_tokens_per_batch_func(
+            pad_token_id=tokenizer.pad_token_id)
+
+    return DataSpec(dataloader=dl, get_num_tokens_in_batch=token_counting_func)
+
+
+def get_tokens_per_batch_func(pad_token_id: int) -> Callable[[Batch], int]:
```

Comment:
As a suggestion, you may want to add a `decoder_only` flag as an input.

I think you can accurately set that flag in all the places where `get_tokens_per_batch_func` gets called. And the resulting function will not have to infer enc-dec from the presence of `decoder_input_ids`. Instead the flag will just determine whether the function should expect `decoder_input_ids` in addition to `input_ids`.

Reply:
ah, yes, I like this idea, will do.